### ISSUE TITLE
Replication wedged with TypeError

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#4705]
Replication got a bit slow, and then there were lots of these in the logs:

```
Traceback (most recent call last):
  File "/home/matrix/synapse/synapse/replication/tcp/protocol.py", line 468, in subscribe_to_stream
    if token > current_token:
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```

This is coming from https://github.com/matrix-org/synapse/blob/a163b748a5ca37853f440c5c46d2da80f738a9e0/synapse/replication/tcp/protocol.py#L468. The problem is that this loop assumes that there are no pending updates with a null `token`; however in practice the pending list may contain batches of updates as per https://github.com/matrix-org/synapse/blob/a163b748a5ca37853f440c5c46d2da80f738a9e0/synapse/replication/tcp/resource.py#L289.

We probably need to unpack the batches to decide whether to send them.

**Comment by anoadragon453** [comment]
So plan for this is:

* if token is `None`, then this is part of a batch and later events might possibly be as well.
* iterate through next events until we reach one that's not `None`
* Take that stream ID, make the check (`> current_token`) if approved send all events from the original to and including the one containing the stream ID.

**Comment by anoadragon453** [comment]
Fixed in #4792
